crates/runtime/src/embeddings/vector_search.rs

```diff
@@ -896,60 +896,75 @@ impl VectorSearch {
                 .get_primary_keys_with_overrides(&self.explicit_primary_keys, &tables)
                 .await?;
 
-            let mut response: VectorSearchResult = HashMap::new();
-            let keywords = keywords.iter().map(std::string::String::as_str).collect::<Vec<&str>>();
+            let keywords = keywords.iter().map(String::as_str).collect::<Vec<&str>>();
 
-            for (tbl, search_vectors) in per_table_embeddings {
-                tracing::debug!("Running vector search for table {:#?}", tbl);
+            let search_futures = per_table_embeddings.into_iter().map(|(tbl, search_vectors)| {
+                let keywords = keywords.clone();
                 let primary_keys = table_primary_keys.get(&tbl).map_or(&[] as &[String], |v| v.as_slice());
 
-                // Only support one embedding column per table.
-                let table_provider = self
-                    .df
-                    .get_table(&tbl)
-                    .await
-                    .ok_or(Error::DataSourcesNotFound {
-                        data_source: vec![tbl.clone()],
-                    })?;
-
-                let Some(embedding_table) = get_embedding_table(&table_provider).await else {
-                    return Err(Error::CannotVectorSearchDataset {
-                        data_source: tbl.clone()
-                    });
-                };
+                async move {
+                    tracing::debug!("Running vector search for table {:#?}", tbl);
+
+                    // Only support one embedding column per table.
+                    let table_provider = self
+                        .df
+                        .get_table(&tbl)
+                        .await
+                        .ok_or(Error::DataSourcesNotFound {
+                            data_source: vec![tbl.clone()],
+                        })?;
+
+                    let Some(embedding_table) = get_embedding_table(&table_provider).await else {
+                        return Err(Error::CannotVectorSearchDataset {
+                            data_source: tbl.clone()
+                        });
+                    };
+
+                    let embedding_columns = embedding_table.get_embedding_columns();
+                    let Some(embedding_column) = embedding_columns.first() else {
+                        return Err(Error::NoEmbeddingColumns {
+                            data_source: tbl.clone(),
+                        });
+                    };
+
+                    if search_vectors.len() != 1 {
+                        return Err(Error::IncorrectNumberOfEmbeddingColumns {
+                            data_source: tbl.clone(),
+                            num_embeddings: search_vectors.len(),
+                        });
+                    }
+                    match search_vectors.first() {
+                        None => unreachable!(),
```

Comment:
```suggestion
                    match search_vectors.first() {
                        None => return Err(Error::IncorrectNumberOfEmbeddingColumns {
                            data_source: tbl.clone(),
                            num_embeddings: search_vectors.len(),
                        }),
```

I know it comes from the existing code, but this length check is redundant - we can just make it the reachable action for a `None` from `.first()`.

Reply:
Thank you for your suggestion!
I have a question before applying it.

```
  if search_vectors.len() != 1 {
      return Err(Error::IncorrectNumberOfEmbeddingColumns {
          data_source: tbl.clone(),
          num_embeddings: search_vectors.len(),
      });
  }
```

Here, `search_vectors.len()` is not allowed not only to be 0, but also 2,3,4... and so on.
But if we removed this checking, then here we can not handle edge case like below

```
  #[snafu(display("Only one embedding column per table currently supported. Table: {} has {num_embeddings} embeddings", data_source.to_string()))]
  IncorrectNumberOfEmbeddingColumns {
      data_source: TableReference,
      num_embeddings: usize,
  },
```

This says "Only one embedding column per table currently supported."
Is it okay not to check it anymore?

Reply:
Ah, yes you're right my bad. In that case, let's just add a note to the unreachable to document this 😄 

```suggestion
                    if search_vectors.len() != 1 {
                        return Err(Error::IncorrectNumberOfEmbeddingColumns {
                            data_source: tbl.clone(),
                            num_embeddings: search_vectors.len(),
                        });
                    }
                    match search_vectors.first() {
                        None => unreachable!("Vector search can only be performed on one embedding column, and the column exists"),
```

Reply:
I have applied your suggestions!